CMakeLists.txt

```
@@ -50,9 +50,11 @@ include(VcPkgManifestFeatures)
 if(NOT VERSION_NUM)
   file(STRINGS "VERSION" VERSION_STR)
   if(NOT "${VERSION_STR}" STREQUAL "")
-    string(REGEX MATCH "([0-9]+\\.[0-9]+\\.[0-9]+)(.*)?" VERSION_PREFIX ${VERSION_STR})
+    string(REGEX MATCH "([0-9]+\\.[0-9]+\\.[0-9]+).*" VERSION_PREFIX ${VERSION_STR})
     set(VERSION_NUM ${CMAKE_MATCH_1})
   endif()
+else()
+  string(REGEX MATCH "[0-9]+\\.[0-9]+\\.[0-9]+.*" VERSION_PREFIX ${VERSION_NUM})

```

no need for capture groups here?

---

In the `NOT VERSION_NUM` case, we end up with something like the following.
```
VERSION_STR = 1.6.0-dev
VERSION_PREFIX = 1.6.0-dev
VERSION_NUM = 1.6.0
```

In the newly added case, we could end up with something like...
```
VERSION_STR = 
VERSION_PREFIX = 1.6.0-dev
VERSION_NUM = 1.6.0-dev
```

I think it needs to be ironed out a bit. Since the user is supplying a value for `VERSION_NUM` directly, maybe we should expect that they will not provide a non-numeric component.

```suggestion
  string(REGEX MATCH "^[0-9]+\\.[0-9]+\\.[0-9]+$" VERSION_PREFIX ${VERSION_NUM})
  if(NOT VERSION_PREFIX)
    message(FATAL_ERROR "Invalid version number: ${VERSION_NUM}")
  endif()
```

If we want, we could also tweak the conditions to check whether the user supplied a value for `VERSION_PREFIX` as well. 🤷

---

VERSION_STR is only used as a temporary variable to hold the file contents, it's never referenced again.

If we want to be more consistent about enforcing the format I reckon we should treat a user supplied VERSION_NUM as if it was provided in VERSION_STR and then do the exact same processing. That's assuming it's even desirable to override the version as a compile time flag anymore. It's an undocumented option not used by any CI builds or referenced in the build instructions so we could remove/rename the expected input variable...

---

I should probably also mention why this matters. The `project()` call uses `VERSION_NUM` directly.

https://github.com/diasurgical/devilutionX/blob/99e12f766362fec90640d3a1e4259d1c0d53ba32/CMakeLists.txt#L77-L79

CMake documentation states that the syntax for the version should be: `<major>[.<minor>[.<patch>[.<tweak>]]]`. If it doesn't match that syntax, the call to `project()` will fail. So it doesn't make sense to allow the regex match to allow version strings we know will fail.

Although, it's worth noting that the current regex would enable the user to supply the `tweak` version by supplying a value like `-DVERSION_NUM=1.6.0.3`.